### PR TITLE
[dns] Add support for static DNS configuration.

### PR DESCRIPTION
**Pull request by oleksandrivantsiv**
#### Why I did it
Add support for static DNS configuration. According to https://github.com/sonic-net/SONiC/pull/1262 HLD.

#### How I did it
Add a new `resolv-config.service` that is responsible for transferring configuration from Config DB into `/etc/resolv.conf` file that is consumed by various subsystems in Linux to resolve domain names into IP addresses.

#### How to verify it
- Run the image compilation. Each component related to the static DNS feature is covered with the unit tests.
- Run sonic-mgmt tests. Static DNS feature will be covered with the system tests.
- Install the image and run manual tests.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211

### CODE REVIEW

**Comment by ganglyu** [comment]
Does this feature cover DNS configuration inside container? For example, "docker exec telemetry cat /etc/resolv.conf"

**Comment by oleksandrivantsiv** [comment]
> Does this feature cover DNS configuration inside container? For example, "docker exec telemetry cat /etc/resolv.conf"

Yes. During the container start the docker daemon copies "/etc/resolv.conf" from the host OS rootfs to the docker container rootfs.
But if the DNS configuration is changed in the run time when the container is running this change won't be reflected inside the docker container. In this case config reload will be required.

**Comment by qiluo-msft** [comment]
> 

`config reload` will impact data plane, which is not necessary for DNS config. Could you improve this PR to update DNS config inside each container?

File: files/image_config/resolv-config/resolv-config.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+WD=/var/run/resolvconf/
+CONFIG_DIR=${WD}/interface/
+STATIC_CONFIG_FILE=mgmt.static
+DYNAMIC_CONFIG_FILE_TEMPLATE=*.dhclient
+
+start()
+{
+    redis-dump -d 4 -k "DNS_NAMESERVER*" -y > /tmp/dns.json
```
Comment (qiluo-msft) [inline files/image_config/resolv-config/resolv-config.sh:10]: >-d 4 [](http://example.com/codeflow?start=15&length=4)

If your intention is to check if the table is empty, no need to dump.

4 is magic number, you may consider `sonic-db-cli` with `CONFIG_DB`.

@judyjoseph to check multi-asic and T2.

**Comment by qiluo-msft** [comment]
> echo 'KUBELET_EXTRA_ARGS="--resolv-conf=/etc/resolv.conf --cgroup-driver=cgroupfs --node-ip=::"' | sudo tee -a  $FILESYSTEM_ROOT/etc/default/kubelet

Wondering if KUBELET_EXTRA_ARGS is impacted by this new feature?

@lixiaoyuner could you help review?

---
Refers to: files/build_templates/sonic_debian_extension.j2:477 in 4fd0180. [](commit_id = 4fd01802a87f2edfe5ac75651ac5e3c7dacf146e, deletion_comment = False)

File: files/image_config/resolv-config/resolv.conf.j2 [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{% for ip in DNS_NAMESERVER|sort %}
+nameserver {{ ip }}
```
Comment (qiluo-msft) [inline files/image_config/resolv-config/resolv.conf.j2:2]: >nameserver [](http://example.com/codeflow?start=0&length=10)

In some use case, people are using "/etc/resolv.conf" to manually config DNS entries. In this PR, the content will be overwrite by j2 rendering.

Could you still allow old use case?

File: src/sonic-config-engine/tests/data/dns/static_dns.json [hunk 3]
```diff
@@ -0,0 +1,6 @@
+{
+    "DNS_NAMESERVER": {
```
Comment (qiluo-msft) [inline src/sonic-config-engine/tests/data/dns/static_dns.json:2]: >DNS_NAMESERVER [](http://example.com/codeflow?start=5&length=14)

Need new yang model for this new ConfigDB table.
Also need yang unit test.

Comment (oleksandrivantsiv) [reply src/sonic-config-engine/tests/data/dns/static_dns.json:2]: Yang model and tests are added: https://github.com/sonic-net/sonic-buildimage/pull/13834

File: files/image_config/resolv-config/resolv.conf.j2 (hunk 2 above)
Comment (oleksandrivantsiv) [reply files/image_config/resolv-config/resolv.conf.j2:2]: Those are mutually exclusive approaches. If the user wants to edit /etc/resolv.conf manually it should not configure DNS via config DB and vice-versa. The default behavior is preserved if no DNS entries are specified in the config DB.

**Pull request closed without merging**